app/javascript/shared/components/plateScanValidators.js

```diff
@@ -130,4 +133,124 @@ const checkQCableWalkingBy = (allowedWalkingByList) => {
+}
+
+const filterWellsWithLibraryCreationRequests = (wells) => {
+  return wells.filter((well) => { return (libraryCreationRequestsFromWell(well).length >= 1) })
+}
+
+const checkMaxCountRequests = (maxWellsWithRequests) => {
+  return (plate) => {
+    const numWellsWithRequest = filterWellsWithLibraryCreationRequests(plate.wells).length
+    if (numWellsWithRequest > maxWellsWithRequests) {
+      return { valid: false, message: 'Plate has more than '+maxWellsWithRequests+' wells with submissions for library preparation ('+numWellsWithRequest+')' }
+    }
+    return validScanMessage()
+  }
+}
+
+const checkMinCountRequests = (minWellsWithRequests) => {
+  return (plate) => {
+    const numWellsWithRequest = filterWellsWithLibraryCreationRequests(plate.wells).length
+    if (numWellsWithRequest < minWellsWithRequests) {
+      return { valid: false, message: 'Plate should have at least '+minWellsWithRequests+' wells with submissions for library preparation ('+numWellsWithRequest+')' }
+    }
+    return validScanMessage()
+  }
+}
+
+const checkAllSamplesInColumnsList = (columnsList) => {
+  return (plate) => {
+    const wells = filterWellsWithLibraryCreationRequests(plate.wells)
+    for (var i=0; i<wells.length; i++) {
+      var well = wells[i]
+      var column = well.position.name.slice(1)
+      if (!columnsList.includes(column)) {
+        return { valid: false, message: 'All samples should be in the columns '+columnsList }
+      }
+    }
+    return validScanMessage()
+  }
+}
+
+const substractArrays = (array_a, array_b) => {
+  return array_a.filter(n => !array_b.includes(n))
+}
+
+const checkLibraryTypesInAllWells = (libraryTypes) => {
+  return (plate) => {
+    for (var posWell=0; posWell<plate.wells.length; posWell++) {
+      var well = plate.wells[posWell]
+      const wellPosition = well.position.name
+      // Only wells with requests are checked
+      const libraryCreationRequests = libraryCreationRequestsFromWell(well)
+      if (libraryCreationRequests.length > 0) {
+        const librariesInWell = libraryCreationRequests.map((request) => request.library_type)
+        const libraryTypesSubstraction = substractArrays(libraryTypes, librariesInWell)
+        if (libraryTypesSubstraction.length != 0) {
+          return { valid: false, 
+            message: 'The well at position '+wellPosition+' is missing libraries: '+libraryTypesSubstraction 
+          }
+        }
+      }
+    }
+    return validScanMessage()
+  }
+}
+
+const getAllSubmissionsWithStateForPlate = (plate, submission_state) => {
+  return filterWellsWithLibraryCreationRequests(plate.wells).map((well) => {
+  //return plate.wells.filter((well) => well.requests_as_source.length > 0 ).map((well) => {
+    return libraryCreationRequestsFromWell(well).filter(
+    //return well.requests_as_source.filter(
+      (request) => request.submission.state == submission_state
+    ).map((request) => request.submission.id)
```

Comment:
We also needed the state to check, is this in reference to the filterProps?

Reply:
Ahh I see, yep, if you do need to access attributes on submission you'll need to include it. However, you're probably better off checking the state of the request instead, as requests can be cancelled or failed independently of the submission. (And if a submission is cancelled, all its requests *should* be cancelled as well) Submission state isn't intended to control workflow and is mostly about the progress of the submission itself. (Cancelling a submission is mostly a convenient way to cancel requests)

Reply:
Although, you don't want to error if you have canceled/failed requests, only if ALL requests are cancelled/failed.

Reply:
The libraryCreationRequestsFromWell is already filtering out any request that is not active; this method ```getAllLibrarySubmissionsWithMatchingStateForPlate ``` in particular is in use in the validation with the submissions for the plates (so it checks all submissions are in the same state), but the handling of canceled and failed requests has already happened before; for example, if a plate has no active requests it will fail because it does not pass the minimum defined of 1 at least.